### PR TITLE
WT-6339 Stop creating snapshots for hs cursors

### PR DESCRIPTION
**Pull request by raviprakashgiri29**
**Stop creating snapshots for history store cursors.**

Earlier the flag `WT_CBT_NO_TXN` was driving create/release snapshots and increment/decrement `session->ncursors`. 
I have added a new flag `WT_CBT_NO_TRACKING` to separate these two operations, now this new flag will drive increment/decrement of `session->ncursors`.

The flag `WT_CBT_NO_TXN` is set in `__wt_hs_cursor_open` not to create snapshot in `__cursor_func_init`. 

FYI - I ran the test/checkpoint binary before/after the changes by enabling statistics and could see a considerable difference in `txn global read lock acquisitions` and `txn global write lock acquisitions` locks

### CODE REVIEW

**Comment by agorrod** [comment]
I'm going to stand down from review, since this idea was mostly mine I'm not very independant :)

**Comment by tammybailey** [comment]
@raviprakashgiri29 it looks like you need to run `s_all`

**Comment by raviprakashgiri29** [comment]
> @raviprakashgiri29 it looks like you need to run `s_all`

@tammybailey : The failure was not because of `s_all` run, one of the comments had the keyword `CBT`. Fixed it.

File: src/include/cursor.i [hunk 1]
```diff
@@ -178,7 +178,7 @@ __cursor_leave(WT_SESSION_IMPL *session)
      * active cursors, and we can release any snapshot we're holding for read committed isolation.
      */
     WT_ASSERT(session, session->ncursors > 0);
-    if (--session->ncursors == 0)
+    if (--session->ncursors == 0 && !F_ISSET(session, WT_CBT_NO_TXN))
```
Comment (ddanderson) [inline src/include/cursor.i:181]: The `WT_CBT_NO_TXN` flag only applies to `WT_CURSOR_BTREE` pointers, it shouldn't apply to session pointers.  Is there any reason this is check needs to change?

Comment (raviprakashgiri29) [reply src/include/cursor.i]: @ddanderson Thanks for explaining to me the background of cbt and session pointers. I have removed the checking from `__cursor_leave` as suggested.

Comment (agorrod) [reply src/include/cursor.i]: I think this check was necessary. We want WT_CBT_NO_TXN to result in cursor tracking happening, but snapshot allocation/free being skipped.
Those two things are correlated here - I don't think __wt_txn_read_last is setup to accept a transaction that doesn't have a snapshot allocated, and skipping the txn_cursor_op call when the WT_CBT_NO_TXN flag means we won't have a snapshot here.

**Comment by ddanderson** [comment]
@raviprakashgiri29, did you run your performance test against the branch with the latest changes, to make sure everything is still good?

File: src/include/cursor.i (hunk 1 above)
Comment (ddanderson) [reply src/include/cursor.i]: @agorrod, given the other changes in this PR, I don't think `__cursor_leave` is ever called with `WT_CBT_NO_TRACKING` on (used to be `WT_CBT_NO_TXN`).  Well, I do see `__cursor_leave` being called in LSM.  Perhaps a cleaner approach is change `__cursor_enter` and `__cursor_leave` to have a cursor argument, and check the new tracking flag within those functions.

**Comment by raviprakashgiri29** [comment]
Earlier `__cursor_enter` was doing only one operation i.e, incrementing ncursors(active cursors) but `__cursor_leave` was doing 2 operations, decrementing the ncursors and releasing the snapshot (tightly coupled). 

I de-coupled decrementing of `ncursors` and releasing the snapshots from the `__cursor_leave` function.

**Comment by raviprakashgiri29** [comment]
> @raviprakashgiri29, did you run your performance test against the branch with the latest changes, to make sure everything is still good?

@ddanderson I ran test/checkpoint binary before/after the change and the stats `txn global read lock acquisitions` and `txn global write lock acquisitions` locks (which gets incremented during creating snapshots) have been reduced.

**Comment by raviprakashgiri29** [comment]
@agorrod Thanks for the discussion. I have de-coupled the releasing of snapshot and decrementing the ncursors from `__cursor_leave` and all the correctness tests are passing after the change.

Can you please review the changes?

File: src/include/cursor.i [hunk 2]
```diff
@@ -198,11 +196,17 @@ __cursor_reset(WT_CURSOR_BTREE *cbt)
...
     if (F_ISSET(cbt, WT_CBT_ACTIVE)) {
-        if (!F_ISSET(cbt, WT_CBT_NO_TXN))
+        if (!F_ISSET(cbt, WT_CBT_NO_TRACKING))
             __cursor_leave(session);
         F_CLR(cbt, WT_CBT_ACTIVE);
     }
 
+    /* When the count of active cursors in the session goes to zero, there are no
+    * active cursors, and we can release any snapshot we're holding for read committed isolation.
+    */
```
Comment (agorrod) [inline src/include/cursor.i:206]: The comment is good, but the style is wrong. There should be a line with only `/*` at the start, followed by the rest of the comment. e.g:

```
/*
 * When the count of active cursors in the session goes to zero, there are no
 * active cursors, and we can release any snapshot we're holding for read committed isolation.
 */
```

File: src/include/cursor.i [hunk 3]
```diff
@@ -174,12 +174,10 @@ static inline void
 __cursor_leave(WT_SESSION_IMPL *session)
 {
     /*
-     * Decrement the count of active cursors in the session. When that goes to zero, there are no
-     * active cursors, and we can release any snapshot we're holding for read committed isolation.
```
Comment (agorrod) [inline src/include/cursor.i:178]: SIngle line comments generally use the form:

```
/* Decrement the count of active cursors in the session. */
```

Similarly above.

File: src/include/cursor.i (hunk 2 above)
Comment (raviprakashgiri29) [reply src/include/cursor.i]: Fixed the comments.

File: src/include/cursor.i (hunk 3 above)
Comment (raviprakashgiri29) [reply src/include/cursor.i:178]: Fixed

**Comment by raviprakashgiri29** [comment]
> LGTM modulo a request for a comment style change.
> Please run this through a MongoDB patch build before merging.

@agorrod Fixed the comment style. Created a MongoDB patch build https://evergreen.mongodb.com/version/5eda3e1c2a60ed0ac4af444e

**Comment by raviprakashgiri29** [comment]
@tammybailey @sulabhM Can you please approve this PR?

**Comment by ddanderson** [comment]
I think there's something odd about this code.

Imagine there are just two cursors left open, one for a regular file (which has neither NO_TXN or NO_TRACKING set) and one for history store (which would have NO_TXN set).  It looks to me like NO_TRACKING is only set for a cursor used by checkpoint.

If I'm reading this right, then if we release the HS cursor first and then the regular cursor, it will result in `__wt_txn_read_last` being called when the regular cursor is closed.  But if we switch the order and release the regular cursor first and then HS cursor, then `__wt_txn_read_last` is not called.

Is that the expected behavior?

**Comment by raviprakashgiri29** [comment]
> I think there's something odd about this code.
> 
> Imagine there are just two cursors left open, one for a regular file (which has neither NO_TXN or NO_TRACKING set) and one for history store (which would have NO_TXN set). It looks to me like NO_TRACKING is only set for a cursor used by checkpoint.
> 
> If I'm reading this right, then if we release the HS cursor first and then the regular cursor, it will result in `__wt_txn_read_last` being called when the regular cursor is closed. But if we switch the order and release the regular cursor first and then HS cursor, then `__wt_txn_read_last` is not called.
> 
> Is that the expected behavior?

@ddanderson Yes, that is correct. As per your example above, we expect history store cursor not to call `__wt_txn_read_last`.

**Comment by ddanderson** [comment]
@raviprakashgiri29 Okay, if that is the intended design, then I agree it meets the design.  It seems odd to me that the final behavior might be different depending on the order that cursors are closed.  @agorrod, could you take a look at my previous comment?  I admit I really probed much into what `__wt_txn_read_last` does.

**Comment by bvpvamsikrishna** [comment]
@raviprakashgiri29 , please create a patch build and link it to this PR.

**Comment by raviprakashgiri29** [comment]
Evergreen patch build link with the latest changes.
https://evergreen.mongodb.com/version/5eda28222a60ed0ac4af1304

**Comment by raviprakashgiri29** [comment]
> @raviprakashgiri29 , please create a patch build and link it to this PR.

@bvpvamsikrishna I have updated the Evergreen patch build link in the Jira ticket.

**Comment by sulabhM** [comment]
Removing myself from review for this one. I think a lot of people looked at it already.

**Comment by agorrod** [comment]
@ddanderson the reason the scenario you outlined should not be a problem is that the history store cursor should be positioned and released within the scope of an operation on the history store. So the ordering you talk about of calling enter and leave in different orders isn't possible.

i.e: It will go like:
cursor_enter (table:blah)
cursor_enter (HistoryStore)
cursor_leave (HistoryStore)
cursor_leave (table:blah)

I've created WT-6389 to encode that restriction and make sure we aren't violating it.

**Comment by raviprakashgiri29** [comment]
Thanks for reviewing the changes.

**Pull request merged**